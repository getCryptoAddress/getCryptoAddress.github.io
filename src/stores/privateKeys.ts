import { defineStore } from "pinia";
import { readonly, ref } from "vue";
import generatePrivateKey from "@/libs/PrivateKeys/generatePrivateKey";
import isValidPrivateKeys from "@/libs/PrivateKeys/isValidPrivateKeys";
import type PrivateKeyFormatted from "@/libs/PrivateKeys/types/PrivateKeyFormatted";
import getFormattedPrivateKey from "@/libs/PrivateKeys/getFormattedPrivateKey";

export const usePrivateKeysStore = defineStore("privateKeys", () => {
  const privateKeys = ref<Uint8Array[]>([]);
  const formattedPrivateKeys = ref<PrivateKeyFormatted[]>();

  async function _generatePrivateKeys() {
    formattedPrivateKeys.value = await Promise.all(
      privateKeys.value.map((key) => getFormattedPrivateKey(key))
    );
  }

  async function setPrivateKeys(keys: (number[] | Uint8Array)[]) {
    const copyKeys = keys.map((key) => new Uint8Array([...key]));

    if (!(await isValidPrivateKeys(copyKeys))) {
      throw new Error("Not valid private keys");
    }
    privateKeys.value = copyKeys;
    await _generatePrivateKeys();
  }

  async function generatePrivatesKeys(count: number) {
    const keys = [];
    for (let i = 0; i < count; i += 1) {
      keys.push(await generatePrivateKey());
    }
    privateKeys.value = keys;
    await _generatePrivateKeys();
  }

  return {
    privateKeys: readonly(privateKeys),
    formattedPrivateKeys: readonly(formattedPrivateKeys),
    generatePrivatesKeys,
    setPrivateKeys,
  };
});
