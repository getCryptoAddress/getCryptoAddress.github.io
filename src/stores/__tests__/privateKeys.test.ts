import { assert, describe, it, expect, beforeEach } from "vitest";
import { setActivePinia, createPinia } from "pinia";
import { usePrivateKeysStore } from "../privateKeys";
import { instantiateSecp256k1 } from "@bitauth/libauth";
import type PrivateKeyFormat from "src/libs/PrivateKeys/types/PrivateKeyFormat";
import { privateKey1, privateKey2, privateKey3 } from "./mocks/keys.mock";

describe("[Store] privateKeys", () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it("privatesKeys generated", async () => {
    const COUNT_KEYS = 51;
    const store = usePrivateKeysStore();
    assert.equal(
      store.privateKeys.length,
      0,
      "Initial store count is not empty"
    );

    await store.generatePrivatesKeys(COUNT_KEYS);
    const uniquePrivatesKeys = [
      ...new Set(store.privateKeys.map((item) => item.join())),
    ];
    assert.equal(
      uniquePrivatesKeys.length,
      COUNT_KEYS,
      `Initial store count is not ${COUNT_KEYS}`
    );

    const secp256k1 = await instantiateSecp256k1();

    assert.isTrue(
      store.privateKeys.every((privateKey) =>
        secp256k1.validatePrivateKey(privateKey)
      ),
      "Not all private keys are valid (wtf?)"
    );

    const expectedFormats: PrivateKeyFormat[] = [
      "uint8Array",
      "array",
      "hex",
      "wifMainnet",
      "wifMainnetUncompressed",
      "wifTestnet",
      "wifTestnetUncompressed",
    ];
    assert.isTrue(
      store.formattedPrivateKeys &&
        store.formattedPrivateKeys.every(
          (privateKey) =>
            JSON.stringify(Object.keys(privateKey).sort()) ===
            JSON.stringify(expectedFormats.sort())
        ),
      "Not all private keys have expected formats"
    );
  });

  it("privateKeys set", async () => {
    const store = usePrivateKeysStore();
    const customKeys = [privateKey1, privateKey2, privateKey3];
    await store.setPrivateKeys(customKeys);
    assert.equal(
      store.privateKeys.length,
      customKeys.length,
      "Initial store count is not empty"
    );

    expect(store.privateKeys).toMatchSnapshot();
    expect(store.formattedPrivateKeys).toMatchSnapshot();
  });
});
