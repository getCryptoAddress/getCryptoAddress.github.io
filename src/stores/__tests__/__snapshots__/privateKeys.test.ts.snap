// Vitest Snapshot v1

exports[`[Store] privateKeys > privateKeys set 1`] = `
[
  Uint8Array [
    130,
    205,
    117,
    44,
    180,
    113,
    245,
    200,
    213,
    19,
    97,
    187,
    70,
    157,
    86,
    111,
    19,
    131,
    14,
    18,
    196,
    57,
    9,
    203,
    64,
    196,
    19,
    26,
    73,
    35,
    126,
    80,
  ],
  Uint8Array [
    201,
    59,
    114,
    215,
    208,
    85,
    88,
    95,
    8,
    25,
    187,
    101,
    2,
    64,
    233,
    47,
    162,
    189,
    25,
    66,
    185,
    4,
    61,
    245,
    90,
    86,
    229,
    154,
    136,
    217,
    111,
    133,
  ],
  Uint8Array [
    2,
    241,
    8,
    120,
    165,
    66,
    56,
    141,
    45,
    212,
    113,
    234,
    18,
    6,
    90,
    33,
    190,
    120,
    117,
    15,
    251,
    178,
    238,
    44,
    141,
    220,
    116,
    78,
    115,
    152,
    67,
    36,
  ],
]
`;

exports[`[Store] privateKeys > privateKeys set 2`] = `
[
  {
    "array": [
      130,
      205,
      117,
      44,
      180,
      113,
      245,
      200,
      213,
      19,
      97,
      187,
      70,
      157,
      86,
      111,
      19,
      131,
      14,
      18,
      196,
      57,
      9,
      203,
      64,
      196,
      19,
      26,
      73,
      35,
      126,
      80,
    ],
    "hex": "82cd752cb471f5c8d51361bb469d566f13830e12c43909cb40c4131a49237e50",
    "uint8Array": Uint8Array [
      130,
      205,
      117,
      44,
      180,
      113,
      245,
      200,
      213,
      19,
      97,
      187,
      70,
      157,
      86,
      111,
      19,
      131,
      14,
      18,
      196,
      57,
      9,
      203,
      64,
      196,
      19,
      26,
      73,
      35,
      126,
      80,
    ],
    "wifMainnet": "L1bySjUcmoxNcdgKpsWiYP4G6McjEgKSDupUxSbmkiRnM1rEU2yH",
    "wifMainnetUncompressed": "5JotktmKXYUxCFCdSM47EuaN6HuW46jh9pK1Shh3tbJrF7yQ91Y",
    "wifTestnet": "cRxxueUUCsedn59bDHKquhZKiav8u8R8Hwxx4s4HFq5nbkvY2vei",
    "wifTestnetUncompressed": "92aXLdas7mZ6AJhv4gx27W8KjxGDDGGtVmAxXL3ZEL3u2AF2Z1B",
  },
  {
    "array": [
      201,
      59,
      114,
      215,
      208,
      85,
      88,
      95,
      8,
      25,
      187,
      101,
      2,
      64,
      233,
      47,
      162,
      189,
      25,
      66,
      185,
      4,
      61,
      245,
      90,
      86,
      229,
      154,
      136,
      217,
      111,
      133,
    ],
    "hex": "c93b72d7d055585f0819bb650240e92fa2bd1942b9043df55a56e59a88d96f85",
    "uint8Array": Uint8Array [
      201,
      59,
      114,
      215,
      208,
      85,
      88,
      95,
      8,
      25,
      187,
      101,
      2,
      64,
      233,
      47,
      162,
      189,
      25,
      66,
      185,
      4,
      61,
      245,
      90,
      86,
      229,
      154,
      136,
      217,
      111,
      133,
    ],
    "wifMainnet": "L3xszekZdGTJmVKN9JfkgVZFM2NxSyhQRA1Hn9SkFjdWFxpMkVDQ",
    "wifMainnetUncompressed": "5KLungpUnNdKXUG359Xz2FB2SP4cfxtkNLG25nsEsvABFps1f8A",
    "wifTestnet": "cUKsTZkR4L9ZvvndXiUt3p4JyFgN7Ro6VC9ktZuFkrHWWhurwkSV",
    "wifTestnetUncompressed": "937YNRe2NbhTVXmKhVRttqiz63RKq8RwiH7yARDkDeuE2oWKjya",
  },
  {
    "array": [
      2,
      241,
      8,
      120,
      165,
      66,
      56,
      141,
      45,
      212,
      113,
      234,
      18,
      6,
      90,
      33,
      190,
      120,
      117,
      15,
      251,
      178,
      238,
      44,
      141,
      220,
      116,
      78,
      115,
      152,
      67,
      36,
    ],
    "hex": "02f10878a542388d2dd471ea12065a21be78750ffbb2ee2c8ddc744e73984324",
    "uint8Array": Uint8Array [
      2,
      241,
      8,
      120,
      165,
      66,
      56,
      141,
      45,
      212,
      113,
      234,
      18,
      6,
      90,
      33,
      190,
      120,
      117,
      15,
      251,
      178,
      238,
      44,
      141,
      220,
      116,
      78,
      115,
      152,
      67,
      36,
    ],
    "wifMainnet": "KwKRpsryq9dPDjGmYfgxW7v9NzChkqChaxk1Qq6JwbdAyHpkeZxP",
    "wifMainnetUncompressed": "5HqaifFx1wWzbbxEGLBsBgBEE8yaFTnFBQxWYgSfxRA6tKij5Hv",
    "wifTestnet": "cMgRHnrqGDKePAk2w5W5sSRD1DW7RHJPeztUXFYpSiHBE2tyCdPr",
    "wifTestnetUncompressed": "91cDJQ5VcAb8ZfTWtg5n4GjBsoLHQdKSXMpTdJoBJ9u9fNFoPvx",
  },
]
`;
